test/hotspot/jtreg/serviceability/dcmd/vm/VMInspectTest.java

```diff
@@ -0,0 +1,269 @@
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class VMInspectTest {
+
+    // - locked <0x00000007dd0135e8> (a MyLock)
+    static Pattern waiting_on_mylock =
+        Pattern.compile("- waiting on \\<0x(\\p{XDigit}+)\\> \\(a MyLock\\)");
+
+    // Event: 2301.131 Thread 0x00007fc6cc2866a0 nmethod 1298 0x00007fc6b4f82610 code [0x00007fc6b4f82a40, 0x00007fc6b4f83aa0]
+    static Pattern compilation_event =
+        Pattern.compile("Event: .* Thread .* nmethod \\d+ 0x(\\p{XDigit}+) code ");
+
+    // Compressed class space mapped at: 0x00007fc62f000000-0x00007fc66f000000, reserved size: 1073741824
+    static Pattern compressed_class_space =
+        Pattern.compile("Compressed class space mapped at: 0x(\\p{XDigit}+)\\-0x(\\p{XDigit}+), ");
+
+    // Narrow klass base: 0x00007fc62e000000, Narrow klass shift: 0, Narrow klass range: 0x100000000
+    static Pattern narrow_klass_base =
+        Pattern.compile("Narrow klass base: 0x(\\p{XDigit}+), ");
+
+    //  tid=0x0000153418029c20
+    static Pattern thread_id_line =
+        Pattern.compile(" tid=0x(\\p{XDigit}+) ");
+
+    public void run(CommandExecutor executor) throws ClassNotFoundException {
+        DcmdTestClass test = new DcmdTestClass();
+        test.work();
+        BigInteger ptr = null;
+        OutputAnalyzer output = null;
+
+        // Testing VM.inspect requires UnlockDiagnosticVMOptions or a debug JVM.
+        // This test runs with a System Property set as a hint whether UnlockDiagnosticVMOptions is set.
+        boolean enabled = Platform.isDebugBuild() || Boolean.getBoolean("vminspect.enabled");
+        System.out.println("VM.inspect should be enabled = " + enabled);
+        if (!enabled) {
+            // Use any pointer, command should be refused:
+            output = executor.execute("VM.inspect 0x0");
+            output.shouldContain("-XX:+UnlockDiagnosticVMOptions is required");
+            return; // no more testing
+        }
+
+        // Tests with VM.inspect enabled:
+        output = executor.execute("help");
+        output.shouldNotContain("VM.inspect"); // VM.inspect is not promoted in help
+        output = executor.execute("help VM.inspect");
+        output.shouldContain("Syntax : VM.inspect"); // but help is available
+
+        testInspectAddress(executor);
+    }
+
+    public void testInspectAddress(CommandExecutor executor) {
+        boolean testMisaligned = true;
+
+        // Test that address is mandatory:
+        // java.lang.IllegalArgumentException: The argument 'address' is mandatory.
+        OutputAnalyzer output = executor.execute("VM.inspect");
+        output.shouldContain("is mandatory");
+
+        // Known bad pointers:
+        output = executor.execute("VM.inspect 0x0");
+        output.shouldContain("address not safe");
+        output = executor.execute("VM.inspect -1");
+        output.shouldContain("address not safe");
+
```

Comment:
Nit: Just a suggestion to make the test more readable. Now when more test cases have been added you may want to refactor it to call a separate method for each sub-test.
E.g.: `testBaddAddresses()`, `testMisalignedAddress()`, `testCompiledMethodAddress()`, `testMetadataAddress()`, `testClassAddress()`, `testThreadAddress()`, etc.